Subclassing context.md (#703)
discord-python/site: pydis_site/apps/content/resources/guides/python-guides/subclassing_context.md

---
title: Subclassing Context

> description: "Subclassing the default `commands.Context` class to add more functionability and customisability."

Reviewer: Nit: customizability *

copy it around or awkwardly import it elsewhere.

This guide will show you how to add a `prompt()` method to the context and how to use it in a command.

## Example subclass and code

The first part - of course - is creating the actual context subclass. This is done similarly to creating a bot
subclass, it will look like this:

```python
from typing import Optional

from discord import RawReactionActionEvent
from discord.ext import commands

class CustomContext(commands.Context):
    async def prompt(
            self,
            message: str,
            *,
            timeout=30.0,
            delete_after=True
    ) -> Optional[bool]:
        """Prompt the author with an interactive confirmation message.

        This method will send the `message` content, and wait for max `timeout` seconds
        (default is `30`) for the author to react to the message.

        If `delete_after` is `True`, the message will be deleted before returning a
        boolean or None (if the author didn't respond) indicating whether the author
        confirmed or denied.
        """
        msg = await self.send(message)

>         for reaction in ('\N{WHITE HEAVY CHECK MARK}', '\N{CROSS MARK}'):

Reviewer 2: I believe this should work and looks better.
```suggestion
        for reaction in ('✅', '❌'):
```

Reviewer 3: This was a conscious decision, as it is generally seen as bad practice to store unicode characters in source files: they are quite difficult to work with. The alternative used here is the most readable one.

That said, I will change this since it appears that I have been outvoted.

            await msg.add_reaction(reaction)

        confirmation = None

>         def check(payload: RawReactionActionEvent):

Reviewer 2: Could we perhaps move this into it's own method? I feel like nested functions and using nonlocal just makes it more confusing and unpythonic.

Reviewer: I agree with Ronin here, if you want your method to be "protected" or "private" then I'd suggest to prefix it with either a "_" or a "__"
Example:

`def _protected_method(...)` **OR** `def __private_method(...)`

Reviewer 3: This function is a closure which closes over `msg`, `confirmation` and `self`. We need access to these during the check.

One possibility *would* be putting this as a method and then using `functools.partial` however I think that a closure is more clear.

                confirmation = True
                return True

            elif emoji == '\N{CROSS MARK}':
                confirmation = False
                return True

            # This means that it was neither of the two emojis added, so the author
            # added some other unrelated reaction.
            return False

        try:
            await self.bot.wait_for('raw_reaction_add', check=check, timeout=timeout)
        except asyncio.TimeoutError:
            # The 'confirmation' variable is still None in this case
            pass

        if delete_after:
            await msg.delete()

        return confirmation
```

After creating your context subclass, you need to override the `get_context()` method on your
Bot class and change the default of the `cls` parameter to this subclass:

```python
from discord.ext import commands

class CustomBot(commands.Bot):
    async def get_context(self, message, *, cls=CustomContext):  # From the above codeblock
        return await super().get_context(message, cls=cls)
```

Now that discord.py is using your custom context, you can use it in a command. For example:

```python
from discord.ext import commands

> bot = CustomBot(...)  # Replace with the arguments for the bot

Reviewer 2: I want to include enabling message content as I see this to be one of the biggest issues currently.
```suggestion
# This makes sure your bot receives message content and run command handlers appropriately.
# Without this, our commands wouldn't work and we wouldn't get any errors either. 
intents = discord.Intents.default()
intents.message_content = True

bot = CustomBot(...)  # Replace with the arguments for the bot
```

@bot.command()

> async def massban(ctx: CustomContext, members: commands.Greedy[discord.Member]):

Reviewer: ```suggestion
async def mass_ban(ctx: CustomContext, members: commands.Greedy[discord.Member]):
```

Reviewer: And I don't see the use of `Greedy` if you have no other parameters after it, so I think a list of discord.Member will work as much

>     if not await ctx.prompt(f"Are you sure you want to ban {len(members)}?"):

Reviewer 2: I think it would be better if we were to extract this part out so it's clearer.
Also the prompt should have "members" at the end, other wise it just looks like:
`Are you sure you want to ban 5?` which doesn't make much sense.

```suggestion
    prompt = await ctx.prompt(f"Are you sure you want to ban {len(members)} members?")
    if not prompt:
```

---
title: Subclassing Context
description: "Subclassing the default commands.Context to add more functionability and customisability."
---

## Basic Subclassing

> First,  a [basic article](https://www.codesdope.com/course/python-subclass-of-a-class/) on subclassing will provide some fundamental knowledge, which is highly suggested before moving on to this topic, as subclassing [Context](https://discordpy.readthedocs.io/en/latest/ext/commands/api.html#discord.ext.commands.Context) can ultimately be a complicated task.

Reviewer 4: I think it's probably better for us to link the official Python docs here (and also trim the whitespace):

```suggestion
First, the [documentation on inheritance](https://docs.python.org/3/tutorial/classes.html#inheritance) on subclassing will provide some fundamental knowledge, which is highly suggested before moving on to this topic, as subclassing [Context](https://discordpy.readthedocs.io/en/latest/ext/commands/api.html#discord.ext.commands.Context) can ultimately be a complicated task.
```

## The benefits of subclassing Context

> Subclassing Context can be very beneficial as it allows you to set custom Context methods that can be used in your code. Some applications and exmpleds are provideded below.

Reviewer 4: ```suggestion
Subclassing Context can be very beneficial as it allows you to set custom Context methods that can be used in your code. Some applications and examples are provided below.
```

> In order to subclass Context, you are required to subclass `commands.Bot` as well. More on subclassing Bot can be read here [link placeholder] but here is a short example.

Reviewer 4: Where will this link go?

```py

> # This View is for the prompt function below in the Context subclass. This is compeltely optional and it's purpose is demonstrating applications of subclassing Context.

Reviewer 4: Typo, let's split to two lines while we're at it:

```suggestion
# This View is for the prompt function below in the Context subclass.
# This is completely optional and its purpose is demonstrating applications of subclassing Context.
```

class PromptView(discord.ui.View):
    def __init__(self, *, timeout: float, author_id: int, ctx: commands.Context, delete_after: bool) -> None:
        super().__init__(timeout=timeout)
        self.value: typing.Optional[bool] = None
        self.delete_after: bool = delete_after
        self.author_id: int = author_id
        self.ctx: Context = ctx
        self.message: typing.Optional[discord.Message] = None

    async def interaction_check(self, interaction: discord.Interaction) -> bool:
        if interaction.user and interaction.user.id == self.author_id:
            return True
        else:
            await interaction.response.send_message('This confirmation dialog is not for you.', ephemeral=True)
            return False

>     # ensure that other members do not confirm or deny

Reviewer 4: As for code style, this could probably be a docstring.

    async def on_timeout(self) -> None:
        if self.delete_after and self.message:
            await self.message.delete()

>         #delete the message if there is no response

Reviewer 4: As for code style, this could probably be a docstring.

    @discord.ui.button(label='Cancel', style=discord.ButtonStyle.red)
    async def cancel(self, button: discord.ui.Button, interaction: discord.Interaction):
        self.value = False #returns False if the cancel button is pressed
        await interaction.response.defer()
        if self.delete_after:
            await interaction.delete_original_message()
        self.stop() #stops the view and deletes the confirmation

# We have made two buttons, one for Confirm and one for Cancel.

class CustomContext(commands.Context):
    # create new Context methods here. For example:

    async def send_embed(self, message, **kwargs):
        return await self.send(embed=discord.Embed(description=message, colour=discord.Color.blurple(), **kwargs))
        # This is a shortcut method to sending a custom embed. Pre-set values can be put here, for example the color kwarg in the example above means it has a pre-set color of blurple

    async def send_error_embed(self, message, **kwargs):
        return await self.send(embed=discord.Embed(description=message, colour=discord.Color.red(), **kwargs))
        # embed with pre-set color of red

>     async def send_sucess_embed(self, message, **kwargs):

Reviewer 4: Typo:

```suggestion
    async def send_success_embed(self, message, **kwargs):
```

        return await self.send(embed=discord.Embed(description=message, colour=discord.Color.green(), **kwargs))
        # embed with pre-set color of green

>     # our custom prompt function using buttons  

Reviewer 4: The comments here are probably better suited to be docstrings, since it's useful for the call site to know about.

>     async def prompt(self,message: str, *, timeout: float = 60.0, delete_after: bool = True, author_id: Optional[int] = None,) -> Optional[bool]:
>         """
>         Parameters
>         message: str
>             The message to show along with the prompt.
> 
>         timeout: float (Defaults to 60 seconds)
>             How long to wait before stopping the view
>         delete_after: bool (Defaults to True)
>             Whether to delete the confirmation message after we're done.
> 
>         author_id: Optional[int] (Defaults to ctx.author)
>             The member who should respond to the prompt.
>         -------
>         Returns  
>             True if the confirm button is pressed,
>             False if cancel button is pressed,
>             None if automatically denied due to no response
>         """

Reviewer 4: This docstring is kind of mangled up. Did you mean this?

```suggestion
    async def prompt(self,message: str, *, timeout: float = 60.0, delete_after: bool = True, author_id: Optional[int] = None,) -> Optional[bool]:
        """
        Parameters
        ----------
        message: str
            The message to show along with the prompt.
        timeout: float (Defaults to 60 seconds)
            How long to wait before stopping the view
        delete_after: bool (Defaults to True)
            Whether to delete the confirmation message after we're done.
        author_id: Optional[int] (Defaults to ctx.author)
            The member who should respond to the prompt.

        Returns  
        -------
        True if the confirm button is pressed,
        False if cancel button is pressed,
        None if automatically denied due to no response
        """
```

        author_id = author_id or self.author.id
        # use the View we have made above and set the arguments
        view = PromptView(
            timeout=timeout,
            delete_after=delete_after,
            ctx=self,
            author_id=author_id,
        )

        view.message = await self.send(message, view=view)
        await view.wait()
        return view.value #Can return True, False or None

class Bot(commands.Bot):
    def __init__(self):
        super().__init__(
            #key word arguments

>         )

Reviewer 4: We're not passing anything here, and I don't think we need to override `__init__` in the first place, do we?

Reviewer 5: I put a comment of `key word arguments` there in the init with the intention of increasing the need to go over the subclassing_bot article to understand how to subclass bot. If I were to add the key word arguments, in my opinion people are more likely to just copy paste without fully understanding (in line 12, I wanted to link the subclassing_bot file). But if this isn't a good reasoning I can change it.

    async def get_context(self, message, *, cls=CustomContext): # When overriding this method, you pass your new Context class to the super() method indicating that there is a new Context class the bot will use.
        return await super().get_context(message, cls=cls)

bot = Bot()

# A command example using subclassed Context:
@bot.command()
async def embed(ctx):
    await ctx.send_embed('This is a subclassed embed!')
    # This would send an embed with the description of the specified string and the embed color wold be Blurple, even though you have not specified it in making the embed, is was set as the pre-set value. You can add more kwargs in sending the embed. For example:

>     await ctx.send_success_embed('This is another example of a subclassed embed!', title='Subclassing Embed', timestamp = datetime.datetime.utcnow())

Reviewer 4: Let's keep code style consistent:

```suggestion
    await ctx.send_success_embed('This is another example of a subclassed embed!', title='Subclassing Embed', timestamp=datetime.datetime.utcnow())
```

# example of the prompt function. It can be used before carrying out big or dangerous operations. For example a massban command where multiple members can be passed in to the member argument
@bot.command()
async def massban(ctx, members : commands.Greedy[discord.Member]):

    request_confirm = await ctx.prompt(f'Are you sure you want to ban {len(members)}?')

>     if request_confirm is False:
>         #the user cancelled so we can return
> 
>     elife request_confirm is True:
>         #the user confirmed so we can continue with the command

Reviewer 4: This is invalid due to `elife`, and we should use more Pythonic checks:

```suggestion
    if request_confirm:
        # The user confirmed so we can continue with the command

    else:
        # The user cancelled so we can return 
```

> ```

Reviewer 4: It may be nice to maybe run the full codeblock through a formatter to adapt it to a single style.

Reviewer 3: I agree, go through this with a linter. It has some inconsistencies and odd-ness.

Please capitalize comments and place them either on the same line, or above, the code.

---

> title: Subclassing Context

Reviewer 3: I am gonna be honest, this guide probably isn't very helpful to most. It only has a missleading example.

I think what we can do, to improve this, is add a full text-explanation of the steps the user has to go through. This can either take the shape of maybe an ordered list with smaller codeblocks of the entire example, or perhaps the full codeblock but with a bunch of comments... or something else I can't think of.

Structuring it like that would be much more helpful to the user!

description: "Subclassing the default commands.Context to add more functionability and customisability."
---

## Basic Subclassing
First, the [documentation on inheritance](https://docs.python.org/3/tutorial/classes.html#inheritance) on subclassing will provide some fundamental knowledge, which is highly suggested before moving on to this topic, as subclassing [Context](https://discordpy.readthedocs.io/en/latest/ext/commands/api.html#discord.ext.commands.Context) can ultimately be a complicated task.

## The benefits of subclassing Context
Subclassing Context can be very beneficial as it allows you to set custom Context methods that can be used in your code. Some applications and examples are provided below.

> In order to subclass Context, you are required to subclass `commands.Bot` as well. More on subclassing Bot can be read here [link placeholder] but here is a short example.

Reviewer 3: Isn't it easier to only mention subclassing the Bot class? You don't need to link to that first documentation, since that is done in the bot one.

> In order to subclass Context, you are required to subclass `commands.Bot` as well. More on subclassing Bot can be read here [link placeholder] but here is a short example.

Reviewer 3: This example, was in-fact, not very short 😅

        else:
            await interaction.response.send_message(
                "This confirmation dialog is not for you.", ephemeral=True
            )
            return False

    async def on_timeout(self) -> None:
        if self.delete_after and self.message:
            await self.message.delete()
        # delete the message if there is no response

    @discord.ui.button(label="Confirm", style=discord.ButtonStyle.green)
    async def confirm(
        self, button: discord.ui.Button, interaction: discord.Interaction
    ):
        self.value = True  # returns True if the confirm button is pressed
        await interaction.response.defer()
        if self.delete_after:
            await interaction.delete_original_message()
        self.stop()  # stops the view and deletes the confirmation message

    @discord.ui.button(label="Cancel", style=discord.ButtonStyle.red)
    async def cancel(self, button: discord.ui.Button, interaction: discord.Interaction):
        self.value = False  # returns False if the cancel button is pressed
        await interaction.response.defer()
        if self.delete_after:
            await interaction.delete_original_message()

>         self.stop()  # stops the view and deletes the confirmation

Reviewer 3: The focus is meant to be on the context, yet this is placed first and is probably longer than the actual context subclass.

Can we consider another type of example?

# We have made two buttons, one for Confirm and one for Cancel.

class CustomContext(commands.Context):
    # create new Context methods here. For example:

    async def send_embed(self, message, **kwargs):
        return await self.send(
            embed=discord.Embed(
                description=message, colour=discord.Color.blurple(), **kwargs
            )
        )

>         # This is a shortcut method to sending a custom embed. Pre-set values can be put here, for example the color kwarg in the example above means it has a pre-set color of blurple

Reviewer 3: This would do great as a docstring.

        # embed with pre-set color of red

    async def send_sucess_embed(self, message, **kwargs):
        return await self.send(
            embed=discord.Embed(
                description=message, colour=discord.Color.green(), **kwargs
            )
        )
        # embed with pre-set color of green

    # our custom prompt function using buttons
    async def prompt(
        self,
        message: str,
        *,
        timeout: float = 60.0,
        delete_after: bool = True,
        author_id: Optional[int] = None,
    ) -> Optional[bool]:
        """
        Parameters
        message: str
            The message to show along with the prompt.

        timeout: float (Defaults to 60 seconds)
            How long to wait before stopping the view
        delete_after: bool (Defaults to True)
            Whether to delete the confirmation message after we're done.

        author_id: Optional[int] (Defaults to ctx.author)
            The member who should respond to the prompt.
        -------
        Returns
            True if the confirm button is pressed,
            False if cancel button is pressed,
            None if automatically denied due to no response

>         """

Reviewer 3: This docstring is missing a description other than the reference of parameters.

Additionally, the typings are present in the definition so you don't need to repeat them in the docstring.

        author_id = author_id or self.author.id
        # use the View we have made above and set the arguments
        view = PromptView(
            timeout=timeout,
            delete_after=delete_after,
            ctx=self,
            author_id=author_id,
        )

        view.message = await self.send(message, view=view)
        await view.wait()
        return view.value  # Can return True, False or None

class Bot(commands.Bot):
    def __init__(self):
        super().__init__(
            # key word arguments
        )

    async def get_context(
        self, message, *, cls=CustomContext
    ):  # When overriding this method, you pass your new Context class to the super() method indicating that there is a new Context class the bot will use.
        return await super().get_context(message, cls=cls)

bot = Bot()

# A command example using subclassed Context:
@bot.command()
async def embed(ctx):
    await ctx.send_embed("This is a custom Context function!")

>     # This would send an embed with the description of the specified string and the embed color wold be Blurple, even though you have not specified it in making the embed, is was set as the pre-set value. You can add more kwargs in sending the embed. For example:

Reviewer 3: This line is waaaaaaaaaaaaaaayyyyyyyy too long

    await ctx.send_success_embed(
        "This is another example of a custom Context function!",
        title="Subclassing Context",
        timestamp=datetime.datetime.utcnow(),
    )

# example of the prompt function. It can be used before carrying out big or dangerous operations. For example a massban command where multiple members can be passed in to the member argument
@bot.command()
async def massban(ctx, members: commands.Greedy[discord.Member]):

    request_confirm = await ctx.prompt(f"Are you sure you want to ban {len(members)}?")
    if request_confirm is False:
        # the user cancelled so we can return
    elif request_confirm is True:

>         # the user confirmed so we can continue with the command

Reviewer 3: This is invalid syntax, if you wish to have the user fill in something please add `...`
